doc/design/design_high-available_sqlflow_server.md

```diff
@@ -0,0 +1,231 @@
+# High-available SQLFlow Server
+
+## Motivations
+
+In the current system, the SQLFlow client connects the SQLFlow server with a long live connection.
+the SQLFlow client sends a gRPC request which contains a SQL statement and wait until the SQLFlow server finishes executing the SQL statement.
+
+When the SQLFlow server receives one training SQL statement, it will generate a python training program that runs on the host or submit the training job to some distributed training service cluster (Kubernetes/Yarn). This will cause:
+
+1. The local job may cause the SQLFlow server resource insufficient when there are too many SQL jobs.
+1. Sometimes, the SQL job takes too much time, and the gRPC calls timeout.
+1. If one of the SQLFlow server instance fails, the SQL job also fails.
+
+In this design, we propose to:
+
+1. Communicate with the SQLFlow server in a polling manner.
+1. Implement `ClusterJobRunner` on the server-side to launch the SQL Job on Kubernetes.
+
+We recommend using `ClusterJobRunner` in the production environment.
+
+## High-Level Design
+
+The high-availabe SQLFlow job workflow is as follows:
+
+<img src="figures/cluster_job_runner.png">
+
+1. SQLFlow client sends the SQL statement via a gRPC call to the SQLFlow server.
+1. For the `LocalJobRunner`:
```

Comment:
I don't see why we need local and server jobs. I understand we need local and Kubernetes job lunchers, which could be a convenient utility to be called by some submitters.

Reply:
For my opinion, `Submitter` and `JobRunner` are two latitude concepts
1. The submitter is a Python program that runs a machine learning job.
1. The jobrunner determinates the SQL job would be run on host or cluster(k8s/Yarn)